File: test/functional/namespace/namespace.test.ts
import {testGeneration} from "../utils.js";

testGeneration("namespace", import.meta.url, output => ({
    input: "lib/**",
    output,
    libraryName: "sandbox-namespace",
    granularity: "top-level",
    packageNameMapper: {
        "will/be/mapped/andthis": "was/mapped/nested",
        "will/be/mapped": "was/mapped/main"
    },
    moduleNameMapper: {
        "will-be-mapped#AndThis": "was-mapped#Nested"
    },
    namespaceStrategy: {
        "package-namespace.ObjectNamespace": "object",
        "package-namespace": "package",
        "IgnoreNamespace": "ignore",
        "will-be-mapped": "package",

        "import-provider": "package",
        "other-import-provider": "package",
        "import-consumer": "package",
    },
    importMapper: {
        "^import-provider": "import.provider",
        "other-import-provider": {
            "default": "other.import.provider.x",
            "\\*": "",
            ".+": "other.import.provider.",
        }
    },
    verbose: true,
}))

File: gradle-plugin/karakum/src/jsTest/functional/io/github/sgrishchenko/karakum/namespace/NamespaceTest.kt
package io.github.sgrishchenko.karakum.namespace

import io.github.sgrishchenko.karakum.configuration.NamespaceStrategy
import io.github.sgrishchenko.karakum.configuration.PartialConfiguration
import io.github.sgrishchenko.karakum.generateTests
import io.github.sgrishchenko.karakum.util.ruleOf
import js.objects.recordOf
import kotlinx.coroutines.test.runTest
import kotlin.test.Test

class NamespaceTest {
    @Test
    fun test() = runTest {
        generateTests("namespace") { output ->
            recordOf(
                "input" to "**/*.d.ts",
                "output" to output,
                "libraryName" to "sandbox-namespace",
                "granularity" to "top-level",
                "packageNameMapper" to recordOf(
                    "will/be/mapped/andthis" to "was/mapped/nested",
                    "will/be/mapped" to "was/mapped/main"
                ),
                "moduleNameMapper" to recordOf(
                    "will-be-mapped#AndThis" to "was-mapped#Nested"
                ),
                "namespaceStrategy" to recordOf(
                    "package-namespace.ObjectNamespace" to NamespaceStrategy.`object`,
                    "package-namespace" to NamespaceStrategy.`package`,
                    "IgnoreNamespace" to NamespaceStrategy.ignore,
                    "will-be-mapped" to NamespaceStrategy.`package`,

                    "import-provider" to NamespaceStrategy.`package`,
                    "other-import-provider" to NamespaceStrategy.`package`,
                    "import-consumer" to NamespaceStrategy.`package`,
                ),
                "importMapper" to recordOf(
                    "^import-provider" to ruleOf("import.provider"),
                    "other-import-provider" to ruleOf(
                        "default" to "other.import.provider.X",
                        "\\*" to "",
                        ".+" to "other.import.provider.",
                    )
                ),
                "verbose" to true,
            ).unsafeCast<PartialConfiguration>()
        }
    }
}
